### DISCUSSION TITLE
Applying a non-linear prestress in a 2D surface

### DISCUSSION THREAD

**Discussion by MusannaGalib** [idaholab/moose#22936] [Q&A Modules: Solid mechanics] [answered]
Hello,

I want to add a non-linear pre-stress along the y-direction to my phase field model using a tensor mechanics module (like the figure). Can anyone give me an idea of how to do this?
![318860706_686092926407549_2613777422738529769_n(1)](https://user-images.githubusercontent.com/64589428/206805405-47b8900c-c00d-4d4d-8445-4f4acd644dd5.jpg)

**Comment by GiudGiud** [comment, marked as answer]
@laagesen @amjokisaari

**Reply by laagesen** [reply]
So it looks like you are running a dendrite growth simulation with the dendrite growing in the y-direction. Are you wanting to apply a stress to the solid dendrite only, or the solid and liquid both? What is the physical mechanism that is causing the stress?

**Reply by MusannaGalib** [reply]
> So it looks like you are running a dendrite growth simulation with the dendrite growing in the y-direction. Are you wanting to apply a stress to the solid dendrite only, or the solid and liquid both? What is the physical mechanism that is causing the stress?

Hello @laagesen,

I want to add stress on the solid part only. This is a pre-stressed that existed on the system even before the battery operation started. 

I am thinking of adding this as an eigen strain (by converting the stress in terms of strain using a compliance tensor). In the code, sigma_o is that strain. I wanted to apply this in the y-direction. Thats why I put eigen_base = '0.0 1.0 0.0'. However, seems like i am doing something wrong - prefactor is not effecting the eigenstrain for some reason

Also, how can I add an equation for sigma_o rather than a constant value? For example,  y = -0.02x^2+0.7591x+0.0079 


Please give your suggestion on this.
```
[./prefactor]
  type = DerivativeParsedMaterial
  args = ' eta disp_x'
  f_name = eigenstrain_prefactor
  constant_names = 'sigma_o'
  constant_expressions = '-0.4'
  function = '(dex/sqrt(dex^2))*sigma_o*(dex/sqrt(dex^2))*((1+eta)/2)'
  material_property_names = ' dex:=D[eta,disp_x]'
[../]
  [./eigenstrain_b]
  type = ComputeVariableEigenstrain
  base_name = phaseb
  eigen_base = '0.0 1.0 0.0'                            
  args = 'eta disp_x'
  prefactor = eigenstrain_prefactor
  eigenstrain_name = eigenstrain
[../]
```

**Reply by laagesen** [reply]
I'm not exactly sure what the physical effect that the eigenstrain would have here- if this is a solid-liquid system, and you cause the solid to expand further, aren't you just going to displace the liquid phase? I guess if you have a constant-volume system you could compress the liquid, but if the eigenstrain isn't changing it doesn't seem like it would affect the dendrite evolution.

That said, if you want to make sure that the eigenstrain is working as you think it should, I would recommend outputting both these material properties to the Exodus file. You can do this for a material by setting `outputs = exodus` in the material block. Then you can visualize what they are and make sure they are doing what you think they should.

**Reply by MusannaGalib** [reply]
Hello @laagesen and @GiudGiud,

I tried to implement the mentioned code, but it diverged after 94 seconds. I am adding the log file here. Can you please give me some suggestions on this?


[log - Copy.txt](https://github.com/idaholab/moose/files/10485579/log.-.Copy.txt)

My scaling used are -
```
[Variables]              
  [./w]                 
  scaling = 1e+3
  [../]
  [./eta]
  scaling = 1e+3
  [../]
  [./pot]
  scaling = 1e+3
  [../]
  [./disp_x]
  scaling = 1e+3
  [../]
  [./disp_y]
  scaling = 1e+3
  [../]
[]
```
Executioners are -
```
[Preconditioning]                #Reduces the total number of linear iterations
  [./SMP]                        #Single Matrix Preconditioner (SMP) builds one preconditioning matrix.
    type = SMP                   #Single Matrix Preconditioner (SMP)
    full = true                  #To use all off diagonal blocks set
      petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm      121                  preonly       lu           4'
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2
  verbose = True
  solve_type ='Newton'
  l_max_its = 50
  l_tol = 1e-5
  nl_max_its = 50
  nl_rel_tol = 1e-5
  nl_abs_tol = 1e-6
    petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm      31                  preonly      lu          4'
  dt=0.01
  end_time = 400
  line_search = 'none'
[]
```

Secondly, Is `eigen_base = '0.0 1.0 0.0'`   represents `eigen_base = 'x y z'`  ?
                For some reason, the mentioned `prefactor` is not affecting the `eigenstrain_b`. Please give your thoughts on this.  I tried to output the material properties using `outputs = exodus`, but the imapcts are not significant. However, if dont use `prefactor` rather than directly give `eigen_base = '0.0 -0.4 0.0'`, it has significant impact.

```
[./prefactor]
  type = DerivativeParsedMaterial
  args = ' eta disp_x'
  f_name = eigenstrain_prefactor
  constant_names = 'sigma_o'
  constant_expressions = '-0.4'
  function = '(dex/sqrt(dex^2))*sigma_o*(dex/sqrt(dex^2))*((1+eta)/2)'
  material_property_names = ' dex:=D[eta,disp_x]'
[../]
  [./eigenstrain_b]
  type = ComputeVariableEigenstrain
  base_name = phaseb
  eigen_base = '0.0 1.0 0.0'                            
  args = 'eta disp_x'
  prefactor = eigenstrain_prefactor
  eigenstrain_name = eigenstrain
[../]
```

**Reply by laagesen** [reply]
For the convergence issues, you are using a Newton solver, which really needs a full and correct Jacobian to work correctly, but the Jacobian couplings between mechanics and phase-field have not been fully implemented. I would try using PJFNK solver instead, there are many examples in the MOOSE phase-field module where you can see examples of how this is done, for example `moose/modules/phase_field/examples/grain_growth/grain_growth_2D_random.i`. I would also try running without coupled mechanics to make sure it converges without the mechanics part first.

For the question about the prefactor not working as you are expecting, I'm not sure what you mean by not `affecting the eigenstrain_b`. Do you mean that it is not affecting microstructural evolution? The reason I suggested outputting `eigenstrain_b` to the Exodus file is so that you can visualize it and make sure it is non-zero. If it is zero everyone then you would know there was a problem with your `prefactor` material property. You can also output `prefactor` to the Exodus file to visualize it and verify that it looks as you expect it to.
